pkg/ops/rawDiskGeneration.go

```diff
@@ -758,3 +812,97 @@ func (r *RawImage) copyShimOrGrub(target, which string) error {
+	// doable but complex:
+	// mount device into a loop device via syscalls
+	// refresh the partition table
+	// create the disk nodes to map to the partitions
+	// then do it in reverse to cleanup
+	// I tried but the devices didnt appear properly dur to it being in a container
+	// so I went with the easy way
+	out, err := exec.Command("losetup", "-D").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to detach loop devices: %w", err)
+	}
+
+	loopDevice, err := exec.Command("losetup", "-f", "--show", image).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to attach file to loop device: %w", err)
+	}
+
+	// clean loop device, trim spaces and such
+	loopDevice = bytes.TrimSpace(loopDevice)
+
+	// Run kpartx
+	out, err = exec.Command("kpartx", "-av", string(loopDevice)).CombinedOutput()
+	if err != nil {
+		internal.Log.Logger.Error().Str("output", string(out)).Msg("kpartx output")
+		return fmt.Errorf("failed to run kpartx: %w", err)
+	}
+	internal.Log.Logger.Debug().Str("loopDevice", string(loopDevice)).Msg("Attached file to loop device")
+
+	defer func() {
+		// TODO: move this to a function
+		out, err := exec.Command("kpartx", "-dv", string(loopDevice)).CombinedOutput()
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Str("device", string(loopDevice)).Str("out", string(out)).Msg("failed to detach loop device")
+			return
+		}
+		out, err = exec.Command("losetup", "-d", string(loopDevice)).CombinedOutput()
+		if err != nil {
+			internal.Log.Logger.Error().Err(err).Str("device", string(loopDevice)).Str("out", string(out)).Msg("failed to detach loop device")
+			return
+		}
+	}()
+
+	internal.Log.Logger.Debug().Str("device", string(loopDevice)).Str("image", image).Msg("Attached image to loop device")
+
+	// While grub is installed to disk + bios_boot partition, it still needs to have the config and mod files available
+	// so, the grub files are stored in the recovery partition
+	// TODO: do not hardcode the partition number
+	// Get the partition number associated to COS_RECOVERY
+	err = unix.Mount("/dev/mapper/loop0p3", tmpDirRecovery, "ext2", 0, "")
+	if err != nil {
+		internal.Log.Logger.Error().Err(err).Str("device", string(loopDevice)).Msg("failed to mount recovery partition")
+		return err
+	}
+
+	defer unix.Unmount(tmpDirRecovery, 0)
+
+	// Unfortunately this is the only way to install grub to a disk
+	// I tried to manually do it but grub does way too much stuff
+	// If someone is braver than me, here are some tips.
+	// boot.img goes to the first sector of the disk, sector 0 (512 bytes)
+	// core.img goes to the partition identified by bios_grub flag
+	// boot.img needs to have the LBA where core.img is located
+	// No idea where is that hardcoded
+	// core.img is not provided by grub packages, it has to be generated beforehand
+	args := []string{
+		"--target=i386-pc",
+		"--force", string(loopDevice),
+		fmt.Sprintf("--boot-directory=%s", tmpDirRecovery),
+	}
```

Comment:
I'm a bit confused. This command is not run in a chroot or anything. If I understand correctly, the code above will mount some more partitions from the img file, but we don't point this command to them in any way. If grub-install is only interested in the recovery partition, can't we run this command against that before we pack everything in an img file? I mean, while everything is still just a directory on the build machine. Then we wouldn't need to loop mount etc.

I'm sure I'm missing something :D

Reply:
well grub2-install does a lot of things behind the scenes.

we need the `tmpDirRecovery` which is the recovery partition mounted, as it will copy the proper artifcats into it (all the mod files and fonts and such)
we need to point to the full image as the install target `loopDevice` becuase it will install boot.img into sector 0 of the image and search for a partition in that device that matches the `bios_boot` flag and install the core.img to it.

So you need to have the final full image set up before running this command, you cant really do it before as all three things need to match for a proper grub install to it.

Is that clear?

Reply:
> Is that clear?

kind of. Let's say "clear enough" :D.